src/client/Pipeline.cpp

```diff
@@ -241,18 +271,41 @@ void PipelineImpl::buildForAppendOrRecovery(bool recovery) {
              */
             if (stage != PIPELINE_SETUP_CREATE && stage != PIPELINE_CLOSE
                     && static_cast<int>(nodes.size()) < replication && canAddDatanode) {
-                if (!addDatanodeToPipeline(excludedNodes)) {
+                // Single data node case
+                bool added = false;
+                if (nodes.empty() && useRemoved) {
+                    if (storageID.length()) {
+                        LOG(INFO, "Pipeline: Adding back only datanode %s", removed.formatAddress().c_str());
+                        nodes.push_back(removed);
+                        lb = filesystem->updateBlockForPipeline(*lastBlock);
+                        storageIDs.push_back(storageID);
+                        lb->setPoolId(lastBlock->getPoolId());
+                        lb->setBlockId(lastBlock->getBlockId());
+                        lb->setLocations(nodes);
+                        lb->setStorageIDs(storageIDs);
+                        lb->setNumBytes(lastBlock->getNumBytes());
+                        lb->setOffset(lastBlock->getOffset());
+                        filesystem->updatePipeline(*lastBlock, *lb, nodes, storageIDs);
+                        added = true;
+                    }
+                }
+                if (!added && !addDatanodeToPipeline(excludedNodes)) {
```

Comment:
That is not enough as we would still want to hit the error code which would then be guarded.

Could be reworked in different PR in the future if we want to.